route/v1/types.go

```diff
@@ -270,8 +270,8 @@ type TLSConfig struct {
 	// insecureEdgeTerminationPolicy indicates the desired behavior for insecure connections to a route. While
 	// each router may make its own decisions on which ports to expose, this is normally port 80.
 	//
-	// * Allow - traffic is sent to the server on the insecure port (default)
-	// * Disable - no traffic is allowed on the insecure port.
+	// * Allow - traffic is sent to the server on the insecure port (edge/reencrypt terminations only) (default).
+	// * None - no traffic is allowed on the insecure port.
 	// * Redirect - clients are redirected to the secure port.
 	InsecureEdgeTerminationPolicy InsecureEdgeTerminationPolicyType `json:"insecureEdgeTerminationPolicy,omitempty" protobuf:"bytes,6,opt,name=insecureEdgeTerminationPolicy,casttype=InsecureEdgeTerminationPolicyType"`
```

Comment:
Is there anything validating the valid values here? Worth adding
```go
// +kubebuilder:validation:Enum=Allow;None;Redirect;""
```

Reply:
Also, worth adding `+kubebuilder:validation:XValidation` to validate Allow is set only for edge/reencrypt TLSTerminationType. Examples: https://github.com/openshift/api/blob/master/insights/v1alpha1/types_insights.go#L94-L97

Reply:
Good point, we Miciah brought this up and we were a little confused. Why does https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml#L156-L179 have validation in the CRD?

It does validate the values you explained here, when I try to add Passthrough & Allow via `oc edit route`:
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# routes.route.openshift.io "router-service1" was not valid:
# * spec.tls.termination: Unsupported value: "Allow": supported values: "edge", "passthrough", "reencrypt"
#

```

Does that mean we still need to add these since there is validation somewhere?

Reply:
I guess a question to clarify is: Is https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml generated by this types.go `./hack/update-codegen-crds.sh`? That's something I don't understand completely. This route.crd.yaml looks to be manually generated.

Reply:
@gcs278,

> I guess a question to clarify is: Is https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml generated by this types.go `./hack/update-codegen-crds.sh`? That's something I don't understand completely. This route.crd.yaml looks to be manually generated.

`route.crd.yaml` is generated by generating the CRD from `route/v1/types.go` and then patching the generated CRD using `route.crd.yaml-patch`.  The patch file can express more complex rules than is possible using `kubebuilder:validation:Enum` and the like.  However, now that [CEL](https://kubernetes.io/docs/reference/using-api/cel/) is available via the `kubebuilder:validation:XValidation` stanza, we could use that (as long as we don't need to backport this PR to earlier versions of OpenShift that don't support CEL).  

@ingvagabund,

> Is there anything validating the valid values here? Worth adding
> 
> ```go
> // +kubebuilder:validation:Enum=Allow;None;Redirect;""
> ```

The validation cannot be expressed using a simple `kubebuilder:validation:Enum` value because the set of allowed values for this field depends on the value of another field. We use https://github.com/openshift/api/blob/master/route/v1/route.crd.yaml-patch to add validation that cannot be expressed using simple `kubebuilder:validation` tags.  

> Also, worth adding `+kubebuilder:validation:XValidation` to validate Allow is set only for edge/reencrypt TLSTerminationType. Examples: https://github.com/openshift/api/blob/master/insights/v1alpha1/types_insights.go#L94-L97. E.g.:
> 
> ```go
> +kubebuilder:validation:XValidation:rule="((self.termination=='passthrough') && (self.insecureEdgeTerminationPolicy=='Allow'))",message="'Allow' insecure edge terminating policy is not valid for passthrough termination"
> ```

If using CEL rather than patches is required going forward, then we can use CEL.  Do we need to rewrite the existing rules from `route.crd.yaml-patch` using CEL too?

Reply:
I think we should lean towards built in validation rather than patches, so I would say use an enum and CEL validation for new stuff, if possible, migrate existing patches later.

I say this because the patches are rather opaque, where the CEL validation and enums are much more forefront when you are looking at the API code, the likelihood of someone missing that there's a patch is much higher than them missing the CEL or enum validations.

On top of that, CEL provides the ability to customise the error messages, where the patches used do not, and they can be pretty terse at times, CEL generally is a better UX IMO

Reply:
> When was this field added?

3.1.

> If it's an existing field, adding additional validation now (that doesn't match what is in the openshift-apiserver validation) would be a breaking change, so we need to be careful of that

The validation that Grant is adding should match the existing validation:

* https://github.com/openshift/library-go/blob/a704a5774e20501fcc70fed8885034fabc266ad7/pkg/route/validation/validation.go#L288-L320
* https://github.com/openshift/api/blob/ce21e3fe5da20b7e3041cc5bb1814672d460960a/route/v1/route.crd.yaml-patch#L68-L86

Reply:
I've added the CEL validation using kubebuilder, removed the route-crd.yaml-patch CRD validation, and tested it out. I tested by changing the name of the CRD (so the API server won't mess with it), and it throws the error message on Allow & Passthrough together as expected and as same as before the new validation.

Changes in https://github.com/openshift/api/commit/692e512a5e6767df03d3a8b58349e7acdc8b0744

Let me know if this is what you all were thinking.

Reply:
@ingvagabund @JoelSpeed Could you take a look the CEL validation to see if this is what you are looking for?

Reply:
@gcs278 Yes. This is what I had in mind. Thank you.

Checking other api types I already see a lot of occurrence of `kubebuilder:validation:XValidation:rule`. @JoelSpeed I see you started introducing the CEL validation in https://github.com/openshift/api/pull/1297 as well. I wonder whether there's actually a way to test the validation. Given we need a running kube-apiserver instance, extending the current e2es is impractical given the overhead of updating openshift/origin. So it seems we still need to rely on validating the validation by hand.

Reply:
I have manually tested the patch by installing it in a test cluster and modifying the fields by hand to ensure combinations work/don't work as expected, but that seems a bit tedious and fragile.

If we only do testing by hand, it would be great if @Miciah or @ingvagabund could manually test check the CRD for me too to make sure I didn't miss something. Could this testing be done with unit testing and mocking a K8S Client?

Reply:
@ingvagabund @gcs278 Whenever we introduce CEL validations, we ask that users create [integration tests](https://github.com/openshift/api/tree/master/tests#readme) based on the format we have created. It runs an API server and then creates the CRD and runs a number of tests to check the validations.

I have yet to review this PR in depth but can pick that up once @ingvagabund is happy with the API

Reply:
I did not perform a complete API review of all fields. Just `InsecureEdgeTerminationPolicy` field. So as far as this goes this looks good to me.

Reply:
@JoelSpeed What is the relationship of the integration tests such as https://github.com/openshift/api/blob/master/route/v1/stable.route.testsuite.yaml and this script: https://github.com/openshift/api/blob/master/route/v1/test-route-validation.sh

I see that passthrough and allow are validated in the script: https://github.com/openshift/api/blob/1b2161d23365fb5918167b2ba73e90ff80ca1805/route/v1/test-route-validation.sh#L442

Should I still update the integration test https://github.com/openshift/api/blob/master/route/v1/stable.route.testsuite.yaml to test this as well?

Reply:
**Update**: I added the integration tests for termination: passthrough and insecureEdgeTerminationPolicy. There's more validation in https://github.com/openshift/api/blob/master/route/v1/test-route-validation.sh, but I just converted what was relevant to this PR, assuming the rest would be converted elsewhere.

I also forgot the error message for the CEL validation (+1 for unit testing).

Let me know if this is what you all are looking for.

https://github.com/openshift/api/compare/044704b43b0f0fa7d61564aa1ba27148d987a3a3..c6baf1fbfd509d649732b84f714f21c52ff7ffc3

Reply:
Yeah that's the kind of test I was looking for, great!